packages/governance/src/governor/governor.cairo

```diff
@@ -0,0 +1,1041 @@
+        ) -> u256 {
+            let voter = starknet::get_caller_address();
+            self._cast_vote(proposal_id, voter, support, "", Immutable::DEFAULT_PARAMS())
+        }
+
+        /// Cast a vote with a `reason`.
+        ///
+        /// Requirements:
+        ///
+        /// - The proposal must be active.
+        ///
+        /// Emits a `VoteCast` event.
+        fn cast_vote_with_reason(
+            ref self: ComponentState<TContractState>,
+            proposal_id: felt252,
+            support: u8,
+            reason: ByteArray
+        ) -> u256 {
+            let voter = starknet::get_caller_address();
+            self._cast_vote(proposal_id, voter, support, reason, Immutable::DEFAULT_PARAMS())
+        }
+
+        /// Cast a vote with a `reason` and additional serialized `params`.
+        ///
+        /// Requirements:
+        ///
+        /// - The proposal must be active.
+        ///
+        /// Emits a `VoteCast` event if no params are provided.
+        /// Emits a `VoteCastWithParams` event otherwise.
+        fn cast_vote_with_reason_and_params(
+            ref self: ComponentState<TContractState>,
+            proposal_id: felt252,
+            support: u8,
+            reason: ByteArray,
+            params: Span<felt252>
+        ) -> u256 {
+            let voter = starknet::get_caller_address();
+            self._cast_vote(proposal_id, voter, support, reason, params)
+        }
+
+        /// Cast a vote using the `voter`'s signature.
+        ///
+        /// Requirements:
+        ///
+        /// - The proposal must be active.
+        /// - The nonce in the signed message must match the account's current nonce.
+        /// - `voter` must implement `SRC6::is_valid_signature`.
+        /// - `signature` should be valid for the message hash.
+        ///
+        /// Emits a `VoteCast` event.
+        fn cast_vote_by_sig(
+            ref self: ComponentState<TContractState>,
+            proposal_id: felt252,
+            support: u8,
+            voter: ContractAddress,
+            signature: Span<felt252>
+        ) -> u256 {
+            // 1. Get and increase current nonce
+            let nonce = self.use_nonce(voter);
+
+            // 2. Build hash for calling `is_valid_signature`
+            let verifying_contract = starknet::get_contract_address();
+            let vote = Vote { verifying_contract, nonce, proposal_id, support, voter };
+            let hash = vote.get_message_hash(voter);
+
+            let is_valid_signature_felt = ISRC6Dispatcher { contract_address: voter }
+                .is_valid_signature(hash, signature.into());
+
+            // 3. Check either 'VALID' or true for backwards compatibility
+            let is_valid_signature = is_valid_signature_felt == starknet::VALIDATED
+                || is_valid_signature_felt == 1;
+
+            assert(is_valid_signature, Errors::INVALID_SIGNATURE);
```

Comment:
This is the third place we use the same boilerplate code to validate a signature, shall we consider extracting it into a util function?

Reply:
Good call. I created a separated [issue](https://github.com/OpenZeppelin/cairo-contracts/issues/1223) for that for the sake of priorities.